### ISSUE TITLE
java.math.BigDecimal.ZERO == 0 returns false

### ISSUE THREAD

**Issue by joeydaowang** [oracle/graal#4677]
**Describe the issue**
We are replacing Nashorn with Graal. Looks all good running our existing JS codes. Just one thing. java.math.BigDecimal.ZERO == 0 returns true in Nashorn but returns false in Graal. I am wondering whether Graal can change behavior here to be aligned with Nashorn, otherwise we need to touch quite a lot JS code for script engine migration.

**Steps to reproduce the issue**
Here is some code `Context.Builder graalContextBuilder = Context.newBuilder()
                .allowExperimentalOptions(true)
                .option("js.nashorn-compat", "true")// make Nashorn external features work e.g. print function
                .allowHostAccess(HostAccess.ALL)
                .allowHostClassLookup((s) -> true);
        GraalJSScriptEngine engine = GraalJSScriptEngine.create(null, graalContextBuilder);
        String script = "var zero = java.math.BigDecimal.ZERO;"
                + " print('zero is 0? ' + (zero == 0));"
                + " print('zero (Number) is 0? ' + (Number(zero) == 0));"
                + " print('zero (Float) is 0? ' + (zero.floatValue() == 0));"
                + " print('zero (Int) is 0? ' + (zero.intValue() == 0));";
        engine.eval(script);`

**Describe GraalVM and your environment:**
 - GraalVM version: **22.1.0**
 - JDK major version: **OpenJDK 11.0.14**
 - OS: **Windows 10**
 - Architecture: **AMD64**

**Comment by iamstolis** [comment]
Thank you for the report. This issue has been fixed recently by https://github.com/oracle/graaljs/commit/4ae2f77c22a37e116d6af13368793d19b2f4b4a0. In other words, `java.math.BigDecimal.ZERO == 0` is `true` in the latest development builds. The fix should be part of [the upcoming](https://www.graalvm.org/release-notes/version-roadmap/) GraalVM 22.2.0.

**Comment by joeydaowang** [comment]
Awesome thanks. 👍

**Comment by joeydaowang** [comment]
With version 22.2.0, java.math.BigDecimal.ZERO == 0 is true. That's great, thanks a lot to fix it 👍 . However java.math.BigDecimal.ZERO === 0 is false, will it be fixed in future version?

**Comment by iamstolis** [comment]
> java.math.BigDecimal.ZERO === 0 is false, will it be fixed in future version?

No, this is the expected result. `===` (identity) is more strict than `==` (equality). `===` is not supposed to perform any conversion and `0` and `java.math.BigDecimal.ZERO` are not identical values. Note that even Nashorn agrees with that:
```
$ jjs
Warning: The jjs tool is planned to be removed from a future JDK release
jjs> java.math.BigDecimal.ZERO == 0
true
jjs> java.math.BigDecimal.ZERO === 0
false
```

**Comment by joeydaowang** [comment]
@iamstolis you are right. Nashorn also has same behavior. We found another piece of code has different behavior in Graal `if(java.math.BigDecimal.ZERO) print('should not be printed');`. With Nashorn, nothing printed. But with Graal, the sentence is printed out. Do you think it is an issue which should be fixed?

**Comment by iamstolis** [comment]
The behavior of `if (something) ...` is based on the conversion of `something` to `boolean`. [The corresponding part of the specification](https://tc39.es/ecma262/#sec-toboolean) says clearly that the result should be `true` if the type of the argument is `object`. So, considering that `typeof java.math.BigDecimal.ZERO === 'object'` (in both Nashorn and GraalJS), it is incorrect (or confusing, at least) that `Boolean(java.math.BigDecimal.ZERO) === false` (in Nashorn). So, I percieve this behavior as a bug/inconsistency in Nashorn. I don't see why `if (java.math.BigDecimal.ZERO) ...` should behave differently than `if (Object(0)) ...`.

**Comment by joeydaowang** [comment]
@iamstolis thanks for the explanation. All good then.

**Comment by joeydaowang** [comment]
@iamstolis another piece of code related to BigDecimal `new java.math.BigDecimal(java.math.BigDecimal.ZERO)` doesn't work in Graal with error **org.graalvm.polyglot.PolyglotException: TypeError: instantiate on java.math.BigDecimal failed due to: no applicable overload found (overloads: [Method[public java.math.BigDecimal(java.math.BigInteger,int)]...**
It works in Nashorn. Will Graal support this kind of expression?

**Comment by iamstolis** [comment]
> It works in Nashorn. Will Graal support this kind of expression?

I am afraid that the answer is no. The method override selection algorithm is more complicated in GraalVM because it must handle more types of objects (objects from other languages, for example). Moreover, the algorithm is more strict when it comes to type conversion (it is more reluctant to perform imprecise/lossy conversions). So, we cannot (and sometimes even don't want to) mimic all subtle details of the method override selection algorithm in Nashorn.

I would argue that in this particular case the right selection is really unclear. Why should be `int`/`long`/`double`/`String` override be preferred over the others? I suggest to rewrite the corresponding code such that it makes the selection more clear (for example, by casting to argument to `Number` or `String` - this will work in both Nashorn and GraalVM then).

Alternatively, you can try to customize the type conversions yourself (using `targetTypeMapping` of `HostAccess`):
```
HostAccess hostAccess = HostAccess.newBuilder(...).targetTypeMapping(BigDecimal.class, Double.class, n -> true, n -> n.doubleValue(), HostAccess.TargetMappingPrecedence.LOWEST)...build();
Context context = Context.newBuilder("js").allowHostAccess(hostAccess)...build()
```
but this approach may not be the best idea for the mentioned example (where the right selection is unclear without an additional hit IMHO).

**Comment by joeydaowang** [comment]
@iamstolis thanks for the explanation and the workaround. All good.
